### PR TITLE
http: Fix UrlEncoding for UTF-8

### PR DESCRIPTION
**Pull request by makunterry**
A quick fix for UTF-8 url encoding support.
Will fix issue:
https://github.com/axel-download-accelerator/axel/issues/345

Signed-off-by: Kun Ma <makunterry@163.com>

### CODE REVIEW

File: src/http.c [hunk 1]
```diff
@@ -427,21 +424,41 @@ void
...
+			}
+
 			j += 2;
 		}
 	}
-	t[j] = 0;
+    t[j] = 0;
 
-	strlcpy(s, t, len);
+    strlcpy(s, t, len);
```
Comment (ismaell) [inline src/http.c:463]: Watch out for inconsistent indentation. Must be tabs.

File: src/http.c [hunk 2]
```diff
@@ -427,21 +424,41 @@ void
...
+				ch == 0x25 /* '%' */ ||
+				ch == 0x5f /* '_' */ ||
+				(ch - 0x2d < 3 && ch >= 0x2d) /* '-','.','/' */ ||
+				(ch - 0x30 < 10 && ch >= 0x30) /* '0' ~ '9' */||
+				(ch - 0x41 < 26 && ch >= 0x41) /* 'A' ~ 'Z' */||
+				(ch - 0x61 < 26 && ch >= 0x61 /* 'a' ~ 'z' */))
+		{
+			t[j] = ch;
+		}
+		else
```
Comment (ismaell) [inline src/http.c:441]: Use the Linux kernel coding style.

File: src/http.c [hunk 3]
```diff
@@ -389,18 +389,15 @@ decode_nibble(char n)
 	return n - 'A' + 10;
 }
 
-inline static char
-encode_nibble(char n)
-{
-	return n > 9 ? n + 'a' - 10 : n + '0';
-}
-
+static unsigned char
```
Comment (ismaell) [inline src/http.c:392]: How is this better?

File: src/http.c [hunk 4]
```diff
@@ -427,21 +424,41 @@ void
...
 	char t[MAX_STRING];
+	char ch = 0;
 	unsigned i, j;
 
-	for (i = j = 0; s[i] && j < sizeof(t) - 1; i++, j++) {
-		t[j] = s[i];
-		if (s[i] <= 0x20 || s[i] >= 0x7f) {
-			/* Fix buffer overflow */
+	for (i = j = 0; i < len && j < sizeof(t) - 1 && (ch = s[i]); i++, j++) {
+		if (ch == 0x2a /* '*' */||
```
Comment (ismaell) [inline src/http.c:431]: too many cases, maybe should be moved to a function

File: src/http.c [hunk 5]
```diff
@@ -427,21 +424,41 @@ void
...
+			t[j] = ch;
+		}
+		else
+		{
 			if (j >= sizeof(t) - 3) {
 				break;
 			}
 
-			encode_byte(t + j, s[i]);
+			if(ch == 0x20)
```
Comment (ismaell) [inline src/http.c:447]: why is %20 a special case?

File: src/http.c (hunk 3 above)
Comment (makunterry) [reply src/http.c:392]: A 16bytes table look up operation should be much more efficient than a ternary operation, generally.

File: src/http.c (hunk 4 above)
Comment (makunterry) [reply src/http.c:431]: The cases are necessary, moving them inside a function can not resolve the complex, but cost 1 more function calling.
I suggest to keep it here, leaving enough comments for each case, will be much easier to maintain.

**Comment by makunterry** [comment]
@ismaell 
Here is testing result, with the same URL from issue https://github.com/axel-download-accelerator/axel/issues/345.
image

File: src/http.c (hunk 4 above)
Comment (ismaell) [reply src/http.c]: There will be no extra cost, the compiler can optimize away the function.

**Comment by ismaell** [comment]
On 04/Apr/2021 06:58, kunm wrote:
> @makunterry commented on this pull request.
> 
> 
> 
> > @@ -389,18 +389,15 @@ decode_nibble(char n)
>  	return n - 'A' + 10;
>  }
>  
> -inline static char
> -encode_nibble(char n)
> -{
> -	return n > 9 ? n + 'a' - 10 : n + '0';
> -}
> -
> +static unsigned char
> 
> A 16bytes table look up operation should be much more efficient than a ternary operation, generally.

Rarely so, in this case it's both slower and bigger.

I just compiled both for x86_64 GCC 10.2:

- original encoder: 32 bytes code
- lookup-table-style encoder: 31 bytes code + 16 bytes data

Not only the machine code takes about the same cache space, it adds a
data load that takes substantial time; compared to that the branch is
insignifficant.

I doubt you can find any modern CPU where the memory is faster than the
branching, and in those cases it would make more sense to produce the
table from the compiler, like they do for jump tables.

File: src/http.c (hunk 3 above)
Comment (ismaell) [reply src/http.c:392]: Please see the reply in the pull request itself.

File: src/http.c [hunk 6]
```diff
@@ -427,17 +424,36 @@ void
...
 	unsigned i, j;
 
-	for (i = j = 0; s[i] && j < sizeof(t) - 1; i++, j++) {
-		t[j] = s[i];
-		if (s[i] <= 0x20 || s[i] >= 0x7f) {
-			/* Fix buffer overflow */
+	for (i = j = 0; i < len && j < sizeof(t) - 1 && (ch = s[i]); i++, j++) {
+		/* There are some charactors in special case that we won't do the encoding:
+		 * 0x20 -> '*'
+		 * 0x25 -> '%'
```
Comment (ismaell) [inline src/http.c:433]: This one can't be skipped.

File: src/http.c (hunk 4 above)
Comment (ismaell) [reply src/http.c]: You can try yourself: https://godbolt.org/z/hc384Gn3d

There's no much difference between `-Os` and `-O2`.

**Comment by makunterry** [comment]
@ismaell 
Thanks for analysis.
And I spent some time on the original code, found that if we use correct char type to do the encoding, everything will be fine.
Please take a look.

**Comment by ismaell** [comment]
After much ruminating over the issue, I conclude there's no case where this is useful, URLs should be always pre-encoded, if anything the code is missing checks to reject malformed URLs.

**Pull request closed without merging**